### PR TITLE
composer set correct numberOfLines on intitial render

### PR DESCRIPTION
**Pull request by Pujan92**
### Details
Update numberOfLines when first time the Composer TextInput renders on the screen and the initial numberOfLines is 0

### Fixed Issues

$ https://github.com/Expensify/App/issues/15024
PROPOSAL: https://github.com/Expensify/App/issues/15024#issuecomment-1428555888

### Tests

1. Open an invalid route(eg. /anything) (Need to test on browser with larger screen)
2. On NotFound page click on close
3. Verify the Composer input not gets distorted

- [X] Verify that no errors appear in the JS console

### Offline tests

1. Open an invalid route(eg. /anything) (Need to test on browser with larger screen)
2. On NotFound page click on close
3. Verify the Composer input not gets distorted

### QA Steps

1. Open an invalid route(eg. /anything) (Need to test on browser with larger screen)
2. On NotFound page click on close
3. Verify the Composer input not gets distorted

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/14358475/220121744-d6020199-2a6e-44e2-82df-f816af21ecf7.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

mobile (8)

</details>

<details>
<summary>Mobile Web - Safari</summary>

mobile (7)

</details>

<details>
<summary>Desktop</summary>

Screenshot 2023-02-20 at 7 07 28 PM

</details>

<details>
<summary>iOS</summary>

Simulator Screen Shot - iPhone 14 - 2023-02-20 at 19 14 45

</details>

<details>
<summary>Android</summary>

Screenshot_1676901602

</details>

### CODE REVIEW

**Comment by marcochavezf** [comment]
Oh interesting, it looks like a bug, do we know why the composer shakes? Is that still happening after applying the solution?

**Comment by Pujan92** [comment]
@thesahindia with your provided scenario I am seeing the rows issue after I did the latest commit with function bind, in the last(Press ➡️ button) step it seems the `scrollHeight` becoming the `0`. Didn't able to find the root cause for it yet but might be calculation gets executed when it still needs to render on the screen(as in the last step ReportScreen gets rerendered).

File: src/components/Composer/index.js [hunk 1]
```diff
@@ -128,6 +128,7 @@ class Composer extends React.Component {
         this.handleWheel = this.handleWheel.bind(this);
         this.putSelectionInClipboard = this.putSelectionInClipboard.bind(this);
         this.shouldCallUpdateNumberOfLines = this.shouldCallUpdateNumberOfLines.bind(this);
+        this.updateNumberOfLinesOnLayoutChange = this.updateNumberOfLinesOnLayoutChange(this);
```
[the remark this answers is not in the document]
Comment (Pujan92) [reply src/components/Composer/index.js]: Oh, how I forgot it. It is working now @thesahindia , updated.

**Comment by marcochavezf** [comment]
Yeah, since the `scrollHeight` will be 0 because the element is not mounted yet, I think we can always return at least `1` for the `numberOfLines` [here](https://github.com/Expensify/App/blob/f95a54e6a24d3d0b860608fc21d4709c0b2d96e0/src/components/Composer/index.js#L347):

```dif
-            const numberOfLines = getNumberOfLines(this.props.maxLines, lineHeight, paddingTopAndBottom, this.textInput.scrollHeight);
+           const numberOfLines = Math.max(getNumberOfLines(this.props.maxLines, lineHeight, paddingTopAndBottom, this.textInput.scrollHeight), 1);
```

And with that, we won't require the function [shouldCallUpdateNumberOfLines](https://github.com/Expensify/App/blob/f95a54e6a24d3d0b860608fc21d4709c0b2d96e0/src/components/Composer/index.js#L321-L329).

@Pujan92 could you please try that approach?

**Comment by marcochavezf** [comment]
> @marcochavezf, there is a problem with that solution. Even when composer has multiline text we will be changing the `numberOfLines` to 1

Oh interesting, what if we use the last `numberOfLines` as the min value?

**Comment by Pujan92** [comment]
> Oh interesting, what if we use the last `numberOfLines` as the min value (instead of 1)?

@marcochavezf that also won't work as this happens for the renders and the rerenders and due to scrollHeight `0` first and last `numberOfLines` might be 0.

**Comment by marcochavezf** [comment]
>that also won't work as this happens for the renders and the rerenders and due to scrollHeight 0 first and last numberOfLines might be 0.

Ah I mean if the text is persisted somehow [here](https://github.com/Expensify/App/pull/15296#issuecomment-1440516203) between renders I think we can also persist the `numberOfLines` somewhere to then use it as a min value.

**Comment by Pujan92** [comment]
> > I think we can also persist the numberOfLines somewhere to then use it as a min value.
> 
> @Pujan92, we are saving the draft message locally. Can you try doing the same for `numberOfLines` and try using it to solve the issue?

@thesahindia Same will be happens to edit message also, won't it increase the keys tremendously in Onyx if we consider every edit comments too? 

https://user-images.githubusercontent.com/14358475/222493136-cb80ae06-c52d-4174-95c4-23d8655adb6f.mov
----
I tried for report composer with `REPORT_DRAFT_COMMENT_NUMBER_OF_LINES: 'reportDraftCommentNumberOfLines_'` and it seems to be working. Can we think of separate PR for this where we are saving the lines in the draft and passing as a props?

**Comment by marcochavezf** [comment]
> We are thinking to store numberOfLines at ONYX to fix the current issue in a more better way so I think it needs to be done in this PR. It's not out of scope.

Yeah I agree it should be done in this PR. And thank you guys for looking for an alternate solution to fix it!

**Comment by Pujan92** [comment]
@thesahindia @marcochavezf with the current implementation I am seeing an issue with the edit message operation(which throws an error for focusing on the composer input ref). It seems to happen with the usage of `withOnyx` in `ReportActionItemMessageEdit` which I added for fetching the `numberOfLines` for the `reportActionID`.

https://user-images.githubusercontent.com/14358475/222896824-adefa51a-f8a2-4ad4-b86c-a8dde02fa7d8.mp4

----
As per my analysis, it(`ReportActionItem`'s `componentDidUpdate`) tries to focus the textInput before it renders. Is the composer rendering delays till the `withOnyx` params do not get fetched asynchronously in parent(I am not sure but the doubt is here for me)? Plz review and check once.

**Comment by Pujan92** [comment]
Two solutions I am seeing for the above issue can be

1. Apply the delay in focus here(currently for animation purpose it is only applied for android)
https://github.com/Expensify/App/blob/f95292aa3facb205df5afb2c57b9a14eac8db61d/src/pages/home/report/ReportActionItem.js#L102

2. Use `initWithStoredValues` to `false` for the onyx mapping [here](https://github.com/Expensify/App/pull/15296/files#diff-c5849fa15c0bb6fc9018f71893e9e91f84308a529b725716ad6d6f0d06ef4089R334) which allow to render immediately and makes the textinput reference available.

@thesahindia @marcochavezf kindly give a look and provide your suggestions, my preference will be option 2.

**Comment by marcochavezf** [comment]
Hi @Pujan92, yes, I also think that `initWithStoredValues: false` is the best option. Thanks!

**Comment by Pujan92** [comment]
@marcochavezf @thesahindia  I proceeded with the `initWithStoredValues: false` change and below is the output of the solution.

https://user-images.githubusercontent.com/14358475/223510726-c4b78808-c6f0-4a49-9095-6e40856a0231.mp4

**Comment by marcochavezf** [comment]
> it looks like changing `initWithStoredValues` to `false` will cause an issue.

I think using the `initWithStoredValues: false` is fine for this case. To solve the issue with the other input, I think we can create another Onyx key for the edited input, something like `REPORT_DRAFT_EDIT_COMMENT_NUMBER_OF_LINES`

**Comment by Pujan92** [comment]
It seems `initWithStoredValues: false` won't work here(for invalid route case) bcoz it will not give the onyx value initially and due to that for an invalid route in the backside we are mounting the `Report` which will set the `numberOfLines` 1 for this edited input. 
@marcochavezf we do have a different pattern for the edited input onyx key so I think another key won't be the solution for it.

**Comment by Pujan92** [comment]
I don't know how to proceed now as I am seeing only an option to delay the focus of the input now, the delay is due to the `numberOfLines` fetching from onyx to render the composer.

**Comment by marcochavezf** [comment]
Yeah, given that the issue [here](https://github.com/Expensify/App/pull/15296#issuecomment-1458630740) is an extreme edge case, we should proceed with the best solution so far 👍🏽

**Comment by Pujan92** [comment]
> @Pujan92, please resolve the conflicts.

@thesahindia resolved.

**Comment by marcochavezf** [comment]
>@marcochavezf, I don't feel good leaving this bug for edit composer while we are fixing the same bug for the main composer. I think we should fix it. Maybe we should open a new https://github.com/Expensify/App/pull/15296#issuecomment-1437508868 for edit composer or maybe we should just move to the original solution where we had this minor issue -

Hi @thesahindia, yeah I agree that we should fix all bugs completely, but given that this was an edge case presented during the PR I think the current solution is ok to continue forward. 

Regarding the solution, I think the solution in this [video](https://github.com/Expensify/App/pull/15296#issuecomment-1458630740) is better than this [one](https://github.com/Expensify/App/pull/15296#issuecomment-1437508868), since from a user perspective the jump looks a bit weirdest than having an edit composer presented in one line (which I think that's not super bad), also seems that @Pujan92 already coded the solution for the jump animation in this PR, no?

**Review by thesahindia — CHANGES_REQUESTED** [review]

File: src/components/Composer/index.js [hunk 2]
```diff
@@ -116,7 +124,7 @@ class Composer extends React.Component {
             : `${props.value || ''}`;
 
         this.state = {
-            numberOfLines: 1,
+            numberOfLines: props.numberOfLines || 1,
```
[the remark this answers is not in the document]
Comment (Pujan92) [reply src/components/Composer/index.js:127]: Yes, thanks.

**Review by thesahindia — APPROVED** [review]
Looks good to me!

cc: @marcochavezf

**Review by marcochavezf — APPROVED** [review]
LGTM, thanks guys!

**Comment by s77rt** [comment]
This PR caused a regression https://github.com/Expensify/App/issues/16931. The use of [`withOnyx` on `numberOfLines`](https://github.com/Pujan92/App/blob/6caefe401fdffe3e24a179789545504b5aa0a6ae/src/pages/home/report/ReportActionItemMessageEdit.js#L334-L339) lead to a partially synced `ReportActionItemMessageEdit` this caused bugs when the component is rendered by two tabs. We should have synced both `numberOfLines` and `draftMessage` or none.

**Pull request merged**